File: supabase/functions/_shared/cors.ts
/**
 * Privxx CORS Headers (Edge Functions)
 * 
 * Restricted CORS for production security.
 * Only allows requests from canonical origin and Lovable preview domains.
 */

// Canonical production origin
const CANONICAL_ORIGIN = 'https://privxx.app';

// Allowed origins pattern
const ALLOWED_ORIGIN_PATTERNS = [
  'https://privxx.app',
  'https://www.privxx.app',
  // Lovable preview domains
  /^https:\/\/[a-z0-9-]+\.lovable\.app$/,
  /^https:\/\/[a-z0-9-]+\.lovableproject\.com$/,
];

/**
 * Check if an origin is allowed
 */
export function isAllowedOrigin(origin: string | null): boolean {
  if (!origin) return false;
  
  return ALLOWED_ORIGIN_PATTERNS.some(pattern => {
    if (typeof pattern === 'string') {
      return origin === pattern;
    }
    return pattern.test(origin);
  });
}

/**
 * Get CORS headers for a request
 * Returns origin-specific headers for allowed origins, or rejects with empty headers
 */
export function getCorsHeaders(requestOrigin: string | null): Record<string, string> {
  // SECURITY: Always validate origin against allowlist - no development bypass
  // This prevents accidental CORS bypass if ENVIRONMENT is misconfigured
  
  if (requestOrigin && isAllowedOrigin(requestOrigin)) {
    return {
      'Access-Control-Allow-Origin': requestOrigin,
      'Access-Control-Allow-Headers': 'authorization, x-client-info, apikey, content-type',
      'Access-Control-Allow-Methods': 'GET, POST, PUT, DELETE, OPTIONS',
      'Access-Control-Allow-Credentials': 'true',
      'Vary': 'Origin',
    };
  }
  
  // Default: return canonical origin (will fail CORS for unauthorized origins)
  return {
    'Access-Control-Allow-Origin': CANONICAL_ORIGIN,
    'Access-Control-Allow-Headers': 'authorization, x-client-info, apikey, content-type',
    'Access-Control-Allow-Methods': 'GET, POST, PUT, DELETE, OPTIONS',
  };
}

/**
 * Handle CORS preflight request
 */
export function handleCorsPreflightResponse(requestOrigin: string | null): Response {
  return new Response(null, {
    status: 204,
    headers: getCorsHeaders(requestOrigin),
  });
}

/**
 * Legacy permissive headers (for backward compatibility during migration)
 * TODO: Remove after verifying production works with strict CORS
 */
export const corsHeaders = {
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Headers': 'authorization, x-client-info, apikey, content-type',
};

File: backend/bridge/cors.go
package main

import (
	"net/http"
	"strings"
)

// allowOrigin returns true if the browser Origin is allowed to call the bridge.
func allowOrigin(origin string) bool {
	// Exact allow-list
	switch origin {
	case "https://privxx.app",
		"https://www.privxx.app",
		"https://privxx.lovable.app":
		return true
	}

	// Suffix allow-list (Lovable preview domains)
	if strings.HasSuffix(origin, ".lovable.app") || strings.HasSuffix(origin, ".lovableproject.com") {
		return true
	}

	return false
}

// cors wraps the handler and:
// - Adds Access-Control-Allow-* for allowed origins
// - Answers preflight OPTIONS with 204 (so the real GET can proceed)
func cors(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		origin := r.Header.Get("Origin")
		if origin != "" && allowOrigin(origin) {
			w.Header().Set("Access-Control-Allow-Origin", origin)
			w.Header().Set("Vary", "Origin")
			w.Header().Set("Access-Control-Allow-Methods", "GET,POST,OPTIONS")
			w.Header().Set("Access-Control-Allow-Headers", "authorization,content-type,apikey,x-request-id,x-user-id")
			w.Header().Set("Access-Control-Allow-Credentials", "false")
		}

		// Preflight: respond here (don’t fall through to /health JSON)
		if r.Method == http.MethodOptions {
			w.WriteHeader(http.StatusNoContent) // 204
			return
		}

		next.ServeHTTP(w, r)
	})
}
